src/bootctl/bootctl-set-efivar.c

```diff
@@ -18,48 +18,74 @@ static int parse_timeout(const char *arg1, char16_t **ret_timeout, size_t *ret_t
         char utf8[DECIMAL_STR_MAX(usec_t)];
         char16_t *encoded;
         usec_t timeout;
-        bool menu_disabled = false;
+        uint64_t loader_features = 0;
+        bool use_arg1 = false;
         int r;
 
         assert(arg1);
         assert(ret_timeout);
         assert(ret_timeout_size);
 
+        assert_cc(STRLEN("menu-force") < ELEMENTSOF(utf8));
+        assert_cc(STRLEN("menu-hidden") < ELEMENTSOF(utf8));
         assert_cc(STRLEN("menu-disabled") < ELEMENTSOF(utf8));
 
-        /* Note: Since there is no way to query if the bootloader supports the string tokens, we explicitly
-         * set their numerical value(s) instead. This means that some of the sd-boot internal ABI has leaked
-         * although the ship has sailed and the side-effects are self-contained.
+        /* Use feature EFI_LOADER_FEATURE_MENU_DISABLE as a mark that the boot loader supports the other
+         * string values too. When unsupported, convert to the timeout with the closest meaning.
          */
-        if (streq(arg1, "menu-force"))
-                timeout = USEC_INFINITY;
-        else if (streq(arg1, "menu-hidden"))
-                timeout = 0;
-        else if (streq(arg1, "menu-disabled")) {
-                uint64_t loader_features = 0;
 
+        if (streq(arg1, "menu-force")) {
                 (void) efi_loader_get_features(&loader_features);
-                if (!(loader_features & EFI_LOADER_FEATURE_MENU_DISABLE)) {
-                        if (arg_graceful() == ARG_GRACEFUL_NO)
-                                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Loader does not support 'menu-disabled'.");
 
-                        log_warning("Loader does not support 'menu-disabled', setting anyway.");
+                if (loader_features & EFI_LOADER_FEATURE_MENU_DISABLE)
+                        use_arg1 = true;
```

Comment:
maybe instead of introducing `use_arg1` just set arg1 to NULL if the numeric timeout value shall be used? and then use arg1 if non-NULL below, and use the numeric value otherwise.